File: implementations/typescript/verify.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Ovidiu Ancuta
//
// aioschema/typescript v0.5.6 | AIOSchema spec v0.5.6
// https://aioschema.org

import * as nodeCrypto from "crypto";
import {
  SUPPORTED_VERSIONS, HASH_REGEX, CORE_HASH_FIELDS, MatchType,
  Manifest, VerificationResult, VerifyOptions, AnchorVerificationError,
  MAX_EXTENSION_SIZE_BYTES
} from "./types";
import {
  computeHash, parseHash, safeEqual, canonicalJson,
  pHashV1, pHashSimilarity,
  SOFT_BINDING_THRESHOLD_DEFAULT, SOFT_BINDING_THRESHOLD_MAX
} from "./algorithms";
import { canonicalManifestBytes } from "./manifest";

function fail(message: string, extra: Partial<VerificationResult> = {}): VerificationResult {
  return {
    success: false, message,
    matchType: "none" as MatchType,
    signatureVerified: false, manifestSignatureVerified: false,
    anchorVerified: false, warnings: [], ...extra
  };
}

function pass(message: string, extra: Partial<VerificationResult> = {}): VerificationResult {
  return {
    success: true, message,
    matchType: "exact" as MatchType,
    signatureVerified: false, manifestSignatureVerified: false,
    anchorVerified: false, warnings: [], ...extra
  };
}

export async function verifyManifest(
  assetData: Uint8Array,
  manifest: Manifest,
  options: VerifyOptions = {}
): Promise<VerificationResult> {
  const warnings: string[] = [];
  const core = manifest.core;
  const ext = manifest.extensions;
  const softThreshold = Math.min(
    options.softThreshold ?? SOFT_BINDING_THRESHOLD_DEFAULT,
    SOFT_BINDING_THRESHOLD_MAX
  );

  // §6.3 — Extension size limit
  if (ext && Object.keys(ext).length > 0) {
    const extSize = Buffer.byteLength(JSON.stringify(ext), "utf-8");
    if (extSize > MAX_EXTENSION_SIZE_BYTES) {
      return fail(`Extensions size (${extSize} bytes) exceeds limit of ${MAX_EXTENSION_SIZE_BYTES} bytes (§6.3)`);
    }
  }

  // §11.1 — ai_declaration constraint validation
  const aiDecl = ext?.ai_declaration as Record<string, unknown> | undefined;
  if (aiDecl && typeof aiDecl === "object") {
    if (aiDecl.standard_editing === true && aiDecl.disclosure_required === true) {
      return fail(
        "ai_declaration constraint violation: standard_editing is true but " +
        "disclosure_required is also true. Per Article 50.2, standard editing " +
        "does not trigger AI disclosure obligations."
      );
    }
    if (aiDecl.human_reviewed === true && !ext?.["compliance_eu_art50"]) {
      warnings.push(
        "ai_declaration.human_reviewed is true but compliance_eu_art50 " +
        "extension is absent (SHOULD be present per §11.1)"
      );
    }
  }

  // §11.3 — public_key fingerprint cross-check
  let embeddedPublicKey: Uint8Array | null = null;
  const pkB64 = ext?.public_key as string | undefined;
  if (pkB64 && typeof pkB64 === "string") {
    try {
      const pkBytes = Buffer.from(pkB64, "base64");
      if (pkBytes.length !== 32) {
        return fail(`extensions.public_key decoded to ${pkBytes.length} bytes, expected 32 (Ed25519)`);
      }
      // Fingerprint cross-check: SHA-256(pubkey)[:32 hex] must match creator_id
      const fpHash = nodeCrypto.createHash("sha256").update(pkBytes).digest("hex").slice(0, 32);
      const expectedCreatorId = `ed25519-fp-${fpHash}`;
      if (!safeEqual(core.creator_id, expectedCreatorId)) {
        return fail(
          "extensions.public_key fingerprint cross-check failed: " +
          "embedded key does not belong to declared creator_id. " +
          `Expected ed25519-fp derived from key: ${expectedCreatorId}, ` +
          `manifest creator_id: ${core.creator_id}`
        );
      }
      embeddedPublicKey = pkBytes;
    } catch {
      return fail("extensions.public_key is not valid Base64");
    }
  }

  // Step 1: schema_version
  if (!SUPPORTED_VERSIONS.has(core.schema_version)) {
    return fail(`Unsupported schema_version '${core.schema_version}'. Supported: [${[...SUPPORTED_VERSIONS].join(", ")}]`);
  }

  // Step 2: required fields
  const missing = (CORE_HASH_FIELDS as readonly string[]).filter(f => !(f in core));
  if (!("core_fingerprint" in core) && !("hash_schema_block" in core)) missing.push("core_fingerprint");
  if (missing.length > 0) return fail(`Missing required core fields: [${missing.sort().join(", ")}]`);

  // Step 3: asset_id
  if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(core.asset_id))
    return fail(`Invalid asset_id format: ${core.asset_id}`);

  // Step 4: creator_id — ed25519-fp- fingerprint or anonymous UUID
  if (!/^ed25519-fp-[0-9a-f]{32}$/.test(core.creator_id) &&
      !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(core.creator_id))
    return fail(`Invalid creator_id format: ${core.creator_id}`);

  // Step 5: hash_original
  const hashes = Array.isArray(core.hash_original) ? core.hash_original : [core.hash_original];
  if (hashes.length === 0) return fail("hash_original must not be empty");
  for (const h of hashes) if (!HASH_REGEX.test(h)) return fail(`Invalid hash_original format: ${h}`);

  // Step 6: core_fingerprint
  const cfp = (core.core_fingerprint ?? core.hash_schema_block) as string;
  if (!cfp || !HASH_REGEX.test(cfp)) return fail(`Invalid core_fingerprint format: ${cfp}`);

  // Step 7: exact hash match
  let exactMatch = false;
  for (const hashStr of hashes) {
    const [alg] = parseHash(hashStr);
    if (safeEqual(computeHash(assetData, alg), hashStr)) { exactMatch = true; break; }
  }

  // Step 8: soft binding
  let matchType: MatchType = "none";
  if (exactMatch) {
    matchType = "exact";
  } else {
    const sb = manifest.extensions?.soft_binding as string | undefined;
    if (sb && typeof sb === "string") {
      const sim = pHashSimilarity(pHashV1(assetData), sb);
      if (sim >= softThreshold) {
        matchType = "soft";
        warnings.push(`Soft binding matched at ${(sim*100).toFixed(1)}%.`);
      }
    }
    if (matchType === "none") {
      return fail("Asset hash does not match manifest.", { matchType, warnings });
    }
  }

  // Step 9: core_fingerprint integrity
  const coreForFp: Record<string, unknown> = {};
  for (const f of CORE_HASH_FIELDS) coreForFp[f] = (core as unknown as Record<string,unknown>)[f];
  const [cfpAlg] = parseHash(cfp);
  if (!safeEqual(computeHash(canonicalJson(coreForFp), cfpAlg), cfp)) {
    return fail("Manifest integrity check failed: core_fingerprint mismatch. Core metadata may have been tampered.", { matchType, warnings });
  }

  // Step 10: signature (optional) — §11.3: prefer embedded key if no external key provided
  const verifyKeyRaw = options.publicKey ?? embeddedPublicKey;
  // Pre-compute SPKI-wrapped public key for signature verification
  const SPKI_ED25519_HEADER = Buffer.from([0x30, 0x2a, 0x30, 0x05, 0x06, 0x03, 0x2b, 0x65, 0x70, 0x03, 0x21, 0x00]);
  let pubKeyObj: ReturnType<typeof nodeCrypto.createPublicKey> | null = null;
  if (verifyKeyRaw) {
    const verifyKeySpki = Buffer.concat([SPKI_ED25519_HEADER, Buffer.from(verifyKeyRaw)]);
    pubKeyObj = nodeCrypto.createPublicKey({ key: verifyKeySpki, format: "der", type: "spki" });
  }

  let signatureVerified = false;
  if (core.signature) {
    if (!verifyKeyRaw) {
      return fail("Manifest is signed but no public key was provided (neither externally nor via extensions.public_key).", { matchType, warnings });
    }
    try {
      const sig = Buffer.from(core.signature.replace(/^ed25519-/, ""), "hex");
      const coreBytes = Buffer.from(canonicalJson(coreForFp), "utf-8");
      signatureVerified = nodeCrypto.verify(null, coreBytes, pubKeyObj!, sig);
      if (!signatureVerified) return fail("Core signature verification failed: invalid signature or wrong key.", { matchType, warnings });
    } catch (e) { return fail(`Signature verification error: ${e instanceof Error ? e.message : String(e)}`, { matchType, warnings }); }
  }

  // Step 11: manifest_signature (optional) — §11.3: prefer embedded key
  let manifestSignatureVerified = false;
  if (core.manifest_signature) {
    if (!verifyKeyRaw) {
      return fail("manifest_signature present but no public key was provided (neither externally nor via extensions.public_key).", { matchType, warnings });
    }
    try {
      // Nullify manifest_signature before computing canonical bytes (§5.8)
      const manifestForSig = { core: { ...manifest.core, manifest_signature: null }, extensions: manifest.extensions };
      const mBytes = canonicalJson(manifestForSig);
      const sig = Buffer.from(core.manifest_signature.replace(/^ed25519-/, ""), "hex");

      manifestSignatureVerified = nodeCrypto.verify(null, mBytes, pubKeyObj, sig);
      if (!manifestSignatureVerified) return fail("Manifest signature verification failed: invalid or extensions tampered.", { matchType, warnings });
    } catch (e) { warnings.push(`Manifest signature verification skipped: crypto error: ${e instanceof Error ? e.message : String(e)}`); }
  }

  // Step 12: anchor (optional)
  let anchorVerified = false;
  if (core.anchor_reference) {
    if (options.verifyAnchor && options.anchorResolver) {
      try {
        const record = await options.anchorResolver(core.anchor_reference);
        if (!record) {
          warnings.push("Anchor record not found.");
        } else {
          const rCfp = (record.core_fingerprint ?? (record as Record<string,unknown>).hash_schema_block ?? "") as string;
          anchorVerified = safeEqual(rCfp, cfp) && safeEqual(record.asset_id, core.asset_id);
          if (!anchorVerified) warnings.push("Anchor record fields do not match manifest.");
        }
      } catch (e) { warnings.push(`Anchor verification failed: ${(e as Error).message}`); }
    } else if (options.verifyAnchor) {
      warnings.push("Anchor present but no anchorResolver provided.");
    } else {
      warnings.push("Anchor present but verifyAnchor=false — not checked.");
    }
  }

  return pass(
    matchType === "exact" ? "✓ PASS: Exact hash match" : "✓ PASS: Soft binding match",
    { matchType, signatureVerified, manifestSignatureVerified, anchorVerified, warnings }
  );
}
// -- end aioschema/typescript v0.5.6 | AIOSchema spec v0.5.6 --

File: implementations/dotnet/Verify.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Ovidiu Ancuta
//
// aioschema/dotnet v0.5.6 | AIOSchema spec v0.5.6
// https://aioschema.org

namespace AIOSchema;

public static class Verifier
{
    public static VerificationResult Verify(
        byte[]? assetBytes,
        AIOSManifest manifest,
        byte[]? publicKey              = null,
        bool verifyAnchor              = false,
        AnchorResolver? anchorResolver = null,
        int softBindingThreshold       = Constants.SoftBindingThresholdDefault)
    {
        var core  = manifest.Core;
        var warns = new List<string>();

        // Step 1 — schema version
        if (!Constants.SupportedVersions.Contains(core.SchemaVersion))
            return VerificationResult.Fail(
                $"Unsupported schema_version '{core.SchemaVersion}'. " +
                $"Supported: {string.Join(", ", Constants.SupportedVersions)}");

        // Step 2 — required fields
        if (string.IsNullOrEmpty(core.AssetId))
            return VerificationResult.Fail("Missing required field: asset_id");
        if (string.IsNullOrEmpty(core.CreationTimestamp))
            return VerificationResult.Fail("Missing required field: creation_timestamp");
        if (string.IsNullOrEmpty(core.CreatorId))
            return VerificationResult.Fail("Missing required field: creator_id");

        var hashes = core.HashOriginalList();
        if (hashes.Count == 0)
            return VerificationResult.Fail("Missing required field: hash_original");

        // §11.1 — ai_declaration constraint + compliance_eu_art50 warning
        if (manifest.Extensions.TryGetValue("ai_declaration", out var aiDeclRaw))
        {
            bool GetBool(object? v) => v switch
            {
                bool b => b,
                System.Text.Json.JsonElement je => je.ValueKind == System.Text.Json.JsonValueKind.True,
                _ => false,
            };

            bool standardEditing = false;
            bool disclosureRequired = false;
            bool humanReviewed = false;

            if (aiDeclRaw is Dictionary<string, object?> adDict)
            {
                adDict.TryGetValue("standard_editing", out var se);
                adDict.TryGetValue("disclosure_required", out var dr);
                adDict.TryGetValue("human_reviewed", out var hr);
                standardEditing = GetBool(se);
                disclosureRequired = GetBool(dr);
                humanReviewed = GetBool(hr);
            }
            else if (aiDeclRaw is System.Text.Json.JsonElement jeDecl &&
                     jeDecl.ValueKind == System.Text.Json.JsonValueKind.Object)
            {
                standardEditing    = jeDecl.TryGetProperty("standard_editing", out var se2)    && se2.GetBoolean();
                disclosureRequired = jeDecl.TryGetProperty("disclosure_required", out var dr2) && dr2.GetBoolean();
                humanReviewed      = jeDecl.TryGetProperty("human_reviewed", out var hr2)      && hr2.GetBoolean();
            }

            if (standardEditing && disclosureRequired)
                return VerificationResult.Fail(
                    "ai_declaration constraint violation: standard_editing is true but " +
                    "disclosure_required is also true. Per Article 50.2, standard editing " +
                    "does not trigger AI disclosure obligations.");

            if (humanReviewed && !manifest.Extensions.ContainsKey("compliance_eu_art50"))
                warns.Add(
                    "ai_declaration.human_reviewed is true but compliance_eu_art50 " +
                    "extension is absent (SHOULD be present per §11.1)");
        }

        // Step 3 — timestamp format
        if (!Patterns.Timestamp().IsMatch(core.CreationTimestamp))
            return VerificationResult.Fail(
                $"creation_timestamp has invalid format '{core.CreationTimestamp}'. " +
                "Must be ISO 8601 UTC with Z suffix.");

        // Step 4 — creator_id format
        try { CreatorId.Parse(core.CreatorId).Validate(); }
        catch (AIOSchemaException ex)
        { return VerificationResult.Fail($"Invalid creator_id: {ex.Message}"); }

        // Step 5 — validate each hash_original value format
        // Note: malformed individual hashes are warned-and-skipped in the hash loop (Step 7).
        // Here we only reject values that are clearly not hash strings at all (no prefix).
        // This matches Go/Rust behavior: warn+skip per entry, fail only if none are supported.

        // Step 6 — optional signature format
        if (core.Signature is not null && !Patterns.Signature().IsMatch(core.Signature))
            return VerificationResult.Fail($"signature has invalid format: '{core.Signature}'");
        if (core.ManifestSignature is not null && !Patterns.Signature().IsMatch(core.ManifestSignature))
            return VerificationResult.Fail($"manifest_signature has invalid format: '{core.ManifestSignature}'");

        // Step 7 — content hash (hard match)
        bool hardMatch     = false;
        bool supportedFound = false;
        string? matchType  = null;

        if (assetBytes is not null)
        {
            foreach (var h in hashes)
            {
                string alg;
                try { (alg, _) = Algorithms.ParseHash(h, "hash_original"); }
                catch (AIOSchemaException ex)
                {
                    warns.Add($"skipping malformed hash '{h}': {ex.Message}");
                    continue;
                }

                if (!Constants.HashRegistry.ContainsKey(alg))
                {
                    warns.Add($"hash algorithm '{alg}' not supported, skipping");
                    continue;
                }

                supportedFound = true;
                if (Algorithms.SafeEqual(Algorithms.ComputeHash(assetBytes, alg), h))
                {
                    hardMatch = true;
                    break;
                }
            }

            if (!supportedFound)
                return VerificationResult.Fail(
                    "No supported hash algorithm found in hash_original. Cannot verify content.");

            // Step 8 — soft binding (pHash requires image processing, not available here)
            if (!hardMatch)
            {
                if (manifest.Extensions.TryGetValue("soft_binding", out var sb)
                    && sb is Dictionary<string, object?> sbDict
                    && sbDict.ContainsKey("fingerprint"))
                {
                    warns.Add(
                        $"soft_binding present but not evaluated " +
                        $"(image processing not available in this implementation; " +
                        $"policy threshold={softBindingThreshold})");
                }
            }

            // Step 9 — fail if no match
            if (!hardMatch)
                return VerificationResult.Fail(
                    "Content mismatch: hash did not match. Asset may be tampered or replaced.");

            matchType = "hard";
        }

        // Step 10 — core_fingerprint integrity (hash_schema_block accepted as alias)
        var cfpValue = core.CoreFingerprint ?? core.HashSchemaBlock;
        if (cfpValue is null)
            return VerificationResult.Fail("Missing required field: core_fingerprint", matchType);

        string cfpAlg;
        try { (cfpAlg, _) = Algorithms.ParseHash(cfpValue, "core_fingerprint"); }
        catch (AIOSchemaException ex)
        { return VerificationResult.Fail(ex.Message, matchType); }

        var coreDict  = ManifestBuilder.CoreToDict(core);
        var canonical = Algorithms.CanonicalCoreBytes(coreDict);
        if (!Algorithms.SafeEqual(Algorithms.ComputeHash(canonical, cfpAlg), cfpValue))
            return VerificationResult.Fail(
                "Manifest integrity check failed: core_fingerprint mismatch. " +
                "Core metadata may have been tampered.", matchType);

        // Step 11 — core signature
        bool signatureVerified = false;
        if (core.Signature is not null)
        {
            if (publicKey is null)
                return VerificationResult.Fail(
                    "Manifest is signed but no public key was provided.", matchType);
            try
            {
                signatureVerified = Algorithms.Verify(canonical, core.Signature, publicKey);
                if (!signatureVerified)
                    return VerificationResult.Fail(
                        "Core signature verification failed: invalid signature or wrong key.", matchType);
            }
            catch (Exception ex)
            { return VerificationResult.Fail($"Signature verification error: {ex.Message}", matchType); }
        }

        // Step 12 — manifest signature
        bool manifestSigVerified = false;
        if (core.ManifestSignature is not null)
        {
            if (publicKey is null)
                return VerificationResult.Fail(
                    "manifest_signature present but no public key was provided.", matchType);
            try
            {
                var mBytes = Algorithms.CanonicalManifestBytes(ManifestBuilder.ToDict(manifest));
                manifestSigVerified = Algorithms.Verify(mBytes, core.ManifestSignature, publicKey);
                if (!manifestSigVerified)
                    return VerificationResult.Fail(
                        "Manifest signature verification failed: invalid or extensions tampered.", matchType);
            }
            catch (Exception ex)
            { return VerificationResult.Fail($"Manifest signature error: {ex.Message}", matchType); }
        }

        // Step 13 — anchor
        bool anchorChecked  = false;
        bool anchorVerified = false;
        if (core.AnchorReference is not null)
        {
            if (verifyAnchor && anchorResolver is not null)
            {
                anchorChecked = true;
                try
                {
                    var rec = anchorResolver(core.AnchorReference);
                    if (rec is null)
                        warns.Add($"Anchor record not found: {core.AnchorReference}");
                    else
                    {
                        var idOk  = rec.TryGetValue("asset_id", out var rId)
                                 && Algorithms.SafeEqual(rId ?? "", core.AssetId);
                        var cfKey = rec.ContainsKey("core_fingerprint") ? "core_fingerprint" : "hash_schema_block";
                        var cfOk  = rec.TryGetValue(cfKey, out var rCfp)
                                 && Algorithms.SafeEqual(rCfp ?? "", cfpValue);
                        if (idOk && cfOk) anchorVerified = true;
                        else warns.Add($"Anchor record mismatch for {core.AnchorReference}.");
                    }
                }
                catch (AnchorVerificationException ex)
                { warns.Add($"Anchor verification error: {ex.Message}"); }
            }
            else
            {
                warns.Add(
                    $"anchor_reference present ({core.AnchorReference}) but not verified. " +
                    "Pass verifyAnchor=true and anchorResolver for Level 3 compliance.");
            }
        }

        // Step 14 — result
        var contentDesc = matchType == "hard" ? "bit-exact"
                        : matchType == "soft"  ? "perceptual (soft)"
                        : "not checked";
        var sigDesc = (signatureVerified && manifestSigVerified) ? "core + manifest signatures verified"
                    : signatureVerified ? "core signature verified"
                    : "unsigned";

        return new VerificationResult
        {
            Success                   = true,
            Message                   = $"Verified: {contentDesc} content match, {sigDesc}. Provenance intact.",
            MatchType                 = matchType,
            SignatureVerified         = signatureVerified,
            ManifestSignatureVerified = manifestSigVerified,
            AnchorChecked             = anchorChecked,
            AnchorVerified            = anchorVerified,
            Warnings                  = warns,
        };
    }
}
// -- end aioschema/dotnet v0.5.6 | AIOSchema spec v0.5.6 --
